Zend/zend_call_stack.c

```diff
@@ -505,6 +509,126 @@ static bool zend_call_stack_get_openbsd(zend_call_stack *stack)
 }
 #endif /* defined(__OpenBSD__) */
 
+#if defined(__NetBSD__)
+# ifdef HAVE_PTHREAD_GETATTR_NP
+static bool zend_call_stack_get_netbsd_pthread(zend_call_stack *stack)
+{
+	pthread_attr_t attr;
+	int error;
+	void *addr;
+	size_t max_size, guard_size;
+
+	error = pthread_getattr_np(pthread_self(), &attr);
+	if (error) {
+		return false;
+	}
+
+	error = pthread_attr_getstack(&attr, &addr, &max_size);
+	if (error) {
+		return false;
+	}
+
+	error = pthread_attr_getguardsize(&attr, &guard_size);
+	if (error) {
+		return false;
+	}
+
+	addr = (char *)addr + guard_size;
+	max_size -= guard_size;
+
+	stack->base = (char *)addr + max_size;
+	stack->max_size = max_size;
+
+	return true;
+}
+# else
+static bool zend_call_stack_get_netbsd_pthread(zend_call_stack *stack)
+{
+	return false;
+}
+# endif /* HAVE_PTHREAD_GETATTR_NP */
+static bool zend_call_stack_get_netbsd_vm(zend_call_stack *stack)
+{
+	/**
+	 * NetBSD supports procfs in a similar fashion as Linux
+	 * however NetBSD's mid/long term plan is to remove it completely.
+	 */
+	char *start, *end;
+	uintptr_t prev_end;
+	struct kinfo_vmentry *entry;
+	char *ebuf;
+	size_t len, max_size;
+	char buffer[4096];
+	uintptr_t addr_on_stack = (uintptr_t)&buffer;
+	int mib[5] = { CTL_VM, VM_PROC, VM_PROC_MAP, getpid(), sizeof(struct kinfo_vmentry) };
+	bool found = false;
+	struct rlimit rlim;
+
+	if (sysctl(mib, 5, NULL, &len, NULL, 0) != 0) {
+		return false;
+	}
+
+	// kinfo_getvmmap uses the same formula, only we do not want to rely on libkvm
+	len = len * 4 / 3 ;
+	ebuf = emalloc(len);
+
+	if (sysctl(mib, 5, ebuf, &len, NULL, 0) != 0) {
+		return false;
+	}
+
+	start = ebuf;
+	end = ebuf + len;
+	prev_end = 0;
+
+	while (start < end) {
+		entry = (struct kinfo_vmentry *)start;
+		if (entry->kve_start <= addr_on_stack && entry->kve_end >= addr_on_stack) {
+			found = true;
+			break;
+		}
+
+		prev_end = entry->kve_end;
+		start += sizeof(struct kinfo_vmentry);
+	}
+
+	if (!found) {
+		efree(ebuf);
+		return false;
+	}
+
+	if (getrlimit(RLIMIT_STACK, &rlim) || rlim.rlim_cur == RLIM_INFINITY) {
+		efree(ebuf);
+		return false;
+	}
+
+	max_size = rlim.rlim_cur;
+
+	if (entry->kve_end - max_size < prev_end) {
+		max_size = prev_end - entry->kve_end;
+	}
+
+	stack->base = (void *)entry->kve_end;
+	stack->max_size = max_size;
+
+	efree(ebuf);
+}
+
+
+static bool zend_call_stack_get_netbsd(zend_call_stack *stack)
+{
+	if (syscall(SYS__lwp_self) == getpid()) {
```

Comment:
`syscall(SYS__lwp_self)` returns `1` on my machine with NetBSD 9.3, and this branch is never entered. A few sources suggest that `_lwp_self()` always returns 1 for the main thread (e.g. [here](http://l.longi.li/blog/2011/02/24/tls-and-shared-library-initialization/)), but I was not able to find evidence in the code.